### ISSUE TITLE
Add support for running separate workers

### ISSUE THREAD

**Issue by bo0tzz** [immich-app/immich-charts#94]
(related: #93)

After the next release, Immich will ship as a single container with the API server and the background workers all combined, with the option to selectively enable only some of those for a running container. Ideally the chart would support running an arbitrary amount of immich containers with configuration for which workers to enable. This would for example enable dedicated transcoding workers with hwacceleration passthrough and such.

**Comment by djjudas21** [comment]
This is an important feature for Kubernetes users. Microservices architecture is best practice for Kube, and running one huge container for everything is wasteful of resources when you have multiple nodes in the cluster.

One of the main things I liked about Immich compared to PhotoPrism is the microservices architecture, so the UI/API doesn't slow down while the app is importing thousands of photos, generating thumbnails, running face recognition, etc.

**Comment by bo0tzz** [comment]
@djjudas21 note that right now there is nothing stopping you from scaling up the single container to multiple replicas. The only tradeoff is you'll be running multiple copies of the api worker as well, but maybe that's a good thing too?

**Comment by djjudas21** [comment]
@bo0tzz I don't know enough about the Immich API - is it safe to run multiple replicas without hitting concurrency issues?

**Comment by bo0tzz** [comment]
Yes, if it wasn't I wouldn't suggest it ;)

**Comment by krypton9208** [comment]
How can i increase number of replicas by helm values for each deployments/StatefulSets ?

**Comment by djjudas21** [comment]
@krypton9208 in case you're still stuck on this, you need something like this in your Helm values to run 2 replicas:

```yaml
server:
  enabled: true
  controller:
    replicas: 2
    strategy: RollingUpdate
```

**Comment by prometheanfire** [comment]
Is the server written in such a way that it can run like this?

**Comment by bo0tzz** [comment]
Yes

**Comment by djjudas21** [comment]
Just to confirm I have been running 2 replicas of the Immich server as per @bo0tzz's recommendation, and it has been working fine for me :ok_hand:

**Comment by daftu** [comment]
Anyone succeeded with autoscaling workers (microservices)?

**Comment by C0nsultant** [comment]
Yes! Check out [this](https://github.com/C0nsultant/homelab-as-code/blob/ff2987a7231067ae5255e256a52335afc04121f7/flux/apps/immich.yaml#L255-L280) pretty stright-forward HPA config based on CPU usage. I've been using it for a while now, it works really well for large ingests. Just make sure you set the degree of parallelism high enough to saturate the resources in whatever your scaling target is.

Of course, this does not _only_ scale the microservices but also the API server. No negative consequence as pointed out by others.

**Comment by daftu** [comment]
I'm trying to modify my fork of this chart but I'm struggling with setting probles for pod with microservices only. I assume it won't work with `/api/server/ping` as I'm disabling api for this containers.
Any suggestions?

**Comment by daftu** [comment]
> I'm trying to modify my fork of this chart but I'm struggling with setting probles for pod with microservices only. I assume it won't work with `/api/server/ping` as I'm disabling api for this containers. Any suggestions?

Managed this by using immich-healtcheck command on "microservices" containers.

**Comment by daftu** [comment]
I've made a PR with separated microservices pod #163

**Comment by ntcong** [comment]
The problem with immich's worker architecture, is that you are either including API (run both worker and API) or excluding (run just worker). So when you're uploading thousands of images and flooding the worker with jobs, the API will still go down.

It's probably better to run more workers AND API, instead of spinning more worker-only pods.

**Comment by xamindar** [comment]
> [@krypton9208](https://github.com/krypton9208) in case you're still stuck on this, you need something like this in your Helm values to run 2 replicas:
> 
> server:
>   enabled: true
>   controller:
>     replicas: 2
>     strategy: RollingUpdate

This doesn't work. The multiple pods will round-robin and pretty much break it. Can't even log in.

**Comment by krypton9208** [comment]
> > [@krypton9208](https://github.com/krypton9208) in case you're still stuck on this, you need something like this in your Helm values to run 2 replicas:
> > server:
> > enabled: true
> > controller:
> > replicas: 2
> > strategy: RollingUpdate
> 
> This doesn't work. The multiple pods will round-robin and pretty much break it. Can't even log in.

Almost year ago I change to docker compose. It works match more efficiently and powerless, I can pass thru gpu by lxc container for encoding… 

For home usages is best options

**Comment by bo0tzz** [comment]
> The multiple pods will round-robin and pretty much break it.

@xamindar what are you seeing exactly? You should be able to run multiple replicas with no issue.

**Comment by bo0tzz** [comment]
@xamindar no idea what you're basing that on, nothing changed around that and it should definitely still work.

**Comment by alphabet5** [comment]
It's not really related to this thread I think, but this is the only place where I found a reference for setting replicas.

The options posted previously did not work, but this did.

```yaml
server:
  enabled: true
  ingress:
    main:
      enabled: false
  controllers:
    main:
      replicas: 2
```

Specifically `controllers` not `controller` and under `main` not directly under `controllers`.

`.server.controllers.main.replicas: 2`
